sqlglot/dialects/bigquery.py

```diff
@@ -933,9 +933,17 @@ def attimezone_sql(self, expression: exp.AtTimeZone) -> str:
             # BigQuery allows CAST(.. AS {STRING|TIMESTAMP} [FORMAT <fmt> [AT TIME ZONE <tz>]]).
             # Only the TIMESTAMP one should use the below conversion, when AT TIME ZONE is included.
             if not isinstance(parent, exp.Cast) or not parent.to.is_type("text"):
-                return self.func(
-                    "TIMESTAMP", self.func("DATETIME", expression.this, expression.args.get("zone"))
-                )
+                from sqlglot.optimizer.annotate_types import annotate_types
+
+                this = annotate_types(expression.this)
```

Comment:
I think we should gate this with `if not this.type`, otherwise we risk overwriting a previously-inferred type.

Reply:
Do we have prior examples of this happening? This looks off to me from an API perspective:
- If type inference is idempotent, this should have the same result so no need for a guard
- If it's not, then the overwrite might have better/newer results than the previous one (?)

I wasn't aware of this so didn't think about adding such a branch, however if it's needed I'd argue it should be baked into the type annotator.

Reply:
This `annotate_types` call has limited context, for example you would infer `x`'s type to be `UNKNOWN`, but if you ran `annotate_types` separately (e.g. what we do in SQLMesh), where you have more context (i.e. schemas), you would know the type beforehand.

Reply:
That makes sense, however I think the guard should still be embedded in the API somehow (e.g. overwrite only if we go from `UNKNOWN` to a proper type), otherwise we'd have to add it across all generator annotations right?

Reply:
We _do_ have it in most, if not all, generator annotations; we should definitely do it everywhere otherwise we could end up in a situation like the one I described above.

> however I think the guard should still be embedded in the API somehow

I don't think this is reasonable behavior. You don't want to make the type inference conditionally overwrite, you want it to always overwrite because otherwise you can't improve previously inferred types. For example if you inferred that an expression has type `INT` with limited context, but then later you have more context (e.g. new/updated schemas) and you can re-infer it to be `BIGINT`, you wanna overwrite the older one there.

Reply:
If it bothers you, you could make a helper like `maybe_annotate_types(E) -> E` which hides the `if expr.type` check.